src/cpp/core/DirectoryBasedFileLock.cpp

```diff
@@ -0,0 +1,222 @@
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <core/FileLock.hpp>
+
+#include <errno.h>
+#include <sys/stat.h>   // ::mkdir
+
+#include <set>
+#include <vector>
+
+#include <core/Thread.hpp>
+#include <core/Error.hpp>
+#include <core/Log.hpp>
+#include <core/FilePath.hpp>
+
+#include <boost/foreach.hpp>
+
+namespace rstudio {
+namespace core {
+
+namespace {
+
+std::string& rstudioLockingPath()
+{
+   static std::string instance("/tmp/rstudio-locking");
+   return instance;
+}
+
+class LockRegistration : boost::noncopyable
+{
+public:
+   
+   void registerLock(const FilePath& lockFilePath)
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         registration_.insert(lockFilePath);
+      }
+      END_LOCK_MUTEX
+   }
+   
+   void deregisterLock(const FilePath& lockFilePath)
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         registration_.erase(lockFilePath);
+      }
+      END_LOCK_MUTEX
+   }
+   
+   void refreshLocks()
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         BOOST_FOREACH(const FilePath& lockFilePath, registration_)
+         {
+            lockFilePath.setLastWriteTime();
+         }
+      }
+      END_LOCK_MUTEX
+   }
+   
+   void clearLocks()
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         BOOST_FOREACH(const FilePath& lockFilePath, registration_)
+         {
+            lockFilePath.removeIfExists();
+         }
+      }
+      END_LOCK_MUTEX
+   }
+   
+   ~LockRegistration()
+   {
+      clearLocks();
+   }
+   
+private:
+   
+   boost::mutex mutex_;
+   std::set<FilePath> registration_;
+   
+};
+
+LockRegistration& lockRegistration()
+{
+   static LockRegistration instance;
+   return instance;
+}
+
+// returns value of 'errno' on failure, or 0 otherwise
+int beginLocking()
+{
+   int status = ::mkdir(rstudioLockingPath().c_str(), 0755);
+   if (status)
+      return errno;
+   return 0;
+}
+
+// returns value of 'errno' on failure, or 0 otherwise
+int endLocking()
+{
+   int status = ::rmdir(rstudioLockingPath().c_str());
+   if (status)
+      return errno;
+   return 0;
+   
+}
+
+struct EndLockingScope : boost::noncopyable
+{
+   ~EndLockingScope()
+   {
+      endLocking();
+   }
+};
+
+} // end anonymous namespace
+
+struct FileLock::Impl
+{
+   FilePath lockFilePath;
+};
+
+FileLock::FileLock()
+   : pImpl_(new Impl())
+{}
+
+FileLock::~FileLock()
+{
+}
+
+FilePath FileLock::lockFilePath() const
+{
+   return pImpl_->lockFilePath;
+}
+
+bool FileLock::isLocked(const FilePath& lockFilePath)
+{
+   if (!lockFilePath.exists())
+      return false;
+   
+   double diff = ::difftime(::time(NULL), lockFilePath.lastWriteTime());
+   return diff < 30.0;
+}
+
+Error FileLock::acquire(const FilePath& lockFilePath)
+{
+   // use filesystem-based mutex to ensure interprocess locking; ie, only
+   // once process can attempt to acquire a lock at a time
+   int errc = beginLocking();
```

Comment:
Since this is using a lockfile approach, this lock should be visible to any machine with access to that filesystem. Is there a place we can write the lock files such that they're visible to all machines? Or do we have an alternative mechanism that could help resolve this?